File: Source/Camping Mod 2.0/common/rikmuld/camping/entity/EntityCamper.java
package rikmuld.camping.entity;

import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Random;

import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.enchantment.EnchantmentThorns;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityAgeable;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.IMerchant;
import net.minecraft.entity.INpc;
import net.minecraft.entity.SharedMonsterAttributes;
import net.minecraft.entity.ai.EntityAIAttackOnCollide;
import net.minecraft.entity.ai.EntityAIMoveTowardsRestriction;
import net.minecraft.entity.ai.EntityAINearestAttackableTarget;
import net.minecraft.entity.ai.EntityAISwimming;
import net.minecraft.entity.ai.EntityAIWander;
import net.minecraft.entity.ai.EntityAIWatchClosest2;
import net.minecraft.entity.monster.EntityMob;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.DamageSource;
import net.minecraft.util.MathHelper;
import net.minecraft.util.Tuple;
import net.minecraft.village.MerchantRecipe;
import net.minecraft.village.MerchantRecipeList;
import net.minecraft.world.World;
import rikmuld.camping.core.register.ModAchievements;
import rikmuld.camping.core.register.ModBlocks;
import rikmuld.camping.core.register.ModItems;
import rikmuld.camping.entity.ai.EntityAILookAtTradePlayerCamper;
import rikmuld.camping.entity.ai.EntityAITradePlayerCamper;
import rikmuld.camping.entity.tileentity.TileEntityCampfireCook;
import rikmuld.camping.item.ItemAnimalStuff;
import rikmuld.camping.item.ItemParts;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

public class EntityCamper extends EntityAgeable implements IMerchant, INpc {

	private boolean isMating;
	private boolean isPlaying;
	Random random = new Random();

	public int[] theCampfire = new int[]{0, 0, 0};
	public int updateHome = 0;

	private EntityPlayer buyingPlayer;

	private MerchantRecipeList buyingList;
	private int timeUntilReset;

	private boolean needsInitilization;
	private int wealth;

	private String lastBuyingPlayer;
	private float field_82191_bN;

	public static final Map buyList = new HashMap();
	public static final Map sellList = new HashMap();

	static
	{
		buyList.put(Integer.valueOf(ModItems.animalStuff.itemID), new Tuple(Integer.valueOf(-4), Integer.valueOf(1)));

		sellList.put(Integer.valueOf(ModItems.hareCooked.itemID), new Tuple(Integer.valueOf(-4), Integer.valueOf(1)));
		sellList.put(Integer.valueOf(ModItems.venisonCooked.itemID), new Tuple(Integer.valueOf(-4), Integer.valueOf(1)));
		sellList.put(Integer.valueOf(ModItems.knife.itemID), new Tuple(Integer.valueOf(1), Integer.valueOf(2)));
		sellList.put(Integer.valueOf(ModItems.backpack.itemID), new Tuple(Integer.valueOf(1), Integer.valueOf(2)));
		sellList.put(Integer.valueOf(ModItems.armorFurLeg.itemID), new Tuple(Integer.valueOf(4), Integer.valueOf(6)));
		sellList.put(Integer.valueOf(ModItems.armorFurHelmet.itemID), new Tuple(Integer.valueOf(3), Integer.valueOf(5)));
		sellList.put(Integer.valueOf(ModItems.armorFurChest.itemID), new Tuple(Integer.valueOf(5), Integer.valueOf(8)));
		sellList.put(Integer.valueOf(ModItems.armorFurBoots.itemID), new Tuple(Integer.valueOf(2), Integer.valueOf(4)));
		sellList.put(Integer.valueOf(ModItems.parts.itemID), new Tuple(Integer.valueOf(1), Integer.valueOf(2)));
	}

	public EntityCamper(World world)
	{
		super(world);
		setGender(random.nextInt(2));
		setSize(0.6F, 1.8F);
		getNavigator().setBreakDoors(true);
		getNavigator().setAvoidsWater(true);
		tasks.addTask(0, new EntityAIMoveTowardsRestriction(this, 1.0F));
		tasks.addTask(1, new EntityAISwimming(this));
		tasks.addTask(2, new EntityAIAttackOnCollide(this, EntityBear.class, 1.0F, false));
		tasks.addTask(2, new EntityAIAttackOnCollide(this, EntityMob.class, 1.0F, false));
		tasks.addTask(2, new EntityAITradePlayerCamper(this));
		tasks.addTask(2, new EntityAILookAtTradePlayerCamper(this));
		tasks.addTask(3, new EntityAIWatchClosest2(this, EntityPlayer.class, 3.0F, 1.0F));
		tasks.addTask(3, new EntityAIWatchClosest2(this, EntityCamper.class, 5.0F, 0.02F));
		tasks.addTask(3, new EntityAIWander(this, 0.6D));
		targetTasks.addTask(0, new EntityAINearestAttackableTarget(this, EntityMob.class, 0, false));
		targetTasks.addTask(0, new EntityAINearestAttackableTarget(this, EntityBear.class, 0, false));
	}

	public EntityCamper(World world, int x, int y, int z)
	{
		this(world);
		setPosition(x, y, z);
		setHomeArea(x, y, z, 8);
	}

	public static void addBlacksmithItem(MerchantRecipeList merchantRecipeList, int id, int meta, Random par2Random, float par3)
	{
		if(par2Random.nextFloat() < par3)
		{
			int j = getRandomCountForBlacksmithItem(id, par2Random);
			ItemStack itemstack;
			ItemStack itemstack1;

			if(j < 0)
			{
				itemstack = new ItemStack(Item.emerald.itemID, 1, 0);
				itemstack1 = new ItemStack(id, -j, meta);
			}
			else
			{
				itemstack = new ItemStack(Item.emerald.itemID, j, 0);
				itemstack1 = new ItemStack(id, 1, meta);
			}

			merchantRecipeList.add(new MerchantRecipe(itemstack, itemstack1));
		}
	}

	public static void addMerchantItem(MerchantRecipeList merchantRecipeList, int id, int meta, Random par2Random, float par3)
	{
		if(par2Random.nextFloat() < par3)
		{
			merchantRecipeList.add(new MerchantRecipe(getRandomSizedStack(id, par2Random, meta), Item.emerald));
		}
	}

	private static int getRandomCountForBlacksmithItem(int par0, Random random)
	{
		Tuple tuple = (Tuple)sellList.get(Integer.valueOf(par0));
		return tuple == null? 1:(((Integer)tuple.getFirst()).intValue() >= ((Integer)tuple.getSecond()).intValue()? ((Integer)tuple.getFirst()).intValue():((Integer)tuple.getFirst()).intValue() + random.nextInt(((Integer)tuple.getSecond()).intValue() - ((Integer)tuple.getFirst()).intValue()));
	}

	private static int getRandomCountForItem(int par0, Random random)
	{
		Tuple tuple = (Tuple)buyList.get(Integer.valueOf(par0));
		return tuple == null? 1:(((Integer)tuple.getFirst()).intValue() >= ((Integer)tuple.getSecond()).intValue()? ((Integer)tuple.getFirst()).intValue():((Integer)tuple.getFirst()).intValue() + random.nextInt(((Integer)tuple.getSecond()).intValue() - ((Integer)tuple.getFirst()).intValue()));
	}

	private static ItemStack getRandomSizedStack(int id, Random random, int meta)
	{
		return new ItemStack(id, getRandomCountForItem(id, random), meta);
	}

	private void addDefaultEquipmentAndRecipies(int par1)
	{
		if(buyingList != null)
		{
			field_82191_bN = MathHelper.sqrt_float(buyingList.size()) * 0.2F;
		}
		else
		{
			field_82191_bN = 0.0F;
		}

		MerchantRecipeList merchantrecipelist;
		merchantrecipelist = new MerchantRecipeList();

		if(getGender() == 0)
		{
			addBlacksmithItem(merchantrecipelist, ModItems.armorFurBoots.itemID, 0, rand, adjustProbability(0.4F));
			addBlacksmithItem(merchantrecipelist, ModItems.armorFurChest.itemID, 0, rand, adjustProbability(0.3F));
			addBlacksmithItem(merchantrecipelist, ModItems.armorFurHelmet.itemID, 0, rand, adjustProbability(0.4F));
			addBlacksmithItem(merchantrecipelist, ModItems.armorFurLeg.itemID, 0, rand, adjustProbability(0.4F));
			addBlacksmithItem(merchantrecipelist, ModItems.knife.itemID, 0, rand, adjustProbability(0.6F));
			addBlacksmithItem(merchantrecipelist, ModItems.backpack.itemID, 0, rand, adjustProbability(0.5F));
		}
		else
		{
			addBlacksmithItem(merchantrecipelist, ModItems.venisonCooked.itemID, 0, rand, adjustProbability(0.6F));
			addBlacksmithItem(merchantrecipelist, ModItems.hareCooked.itemID, 0, rand, adjustProbability(0.6F));
			addBlacksmithItem(merchantrecipelist, ModItems.parts.itemID, ItemParts.PAN, rand, adjustProbability(0.6F));
			addMerchantItem(merchantrecipelist, ModItems.animalStuff.itemID, ItemAnimalStuff.ANTLER, rand, 0.25F);
		}

		if(merchantrecipelist.isEmpty())
		{
			addMerchantItem(merchantrecipelist, ModItems.animalStuff.itemID, ItemAnimalStuff.ANTLER, rand, 1.0F);
		}

		Collections.shuffle(merchantrecipelist);

		if(buyingList == null)
		{
			buyingList = new MerchantRecipeList();
		}

		for(int j1 = 0; (j1 < par1) && (j1 < merchantrecipelist.size()); ++j1)
		{
			buyingList.addToListWithCheck((MerchantRecipe)merchantrecipelist.get(j1));
		}
	}

	private float adjustProbability(float par1)
	{
		float f1 = par1 + field_82191_bN;
		return f1 > 0.9F? 0.9F - (f1 - 0.9F):f1;
	}

	@Override
	public boolean allowLeashing()
	{
		return false;
	}

	@Override
	protected void applyEntityAttributes()
	{
		super.applyEntityAttributes();
		getEntityAttribute(SharedMonsterAttributes.movementSpeed).setAttribute(0.4D);
		getAttributeMap().func_111150_b(SharedMonsterAttributes.attackDamage).setAttribute(4.0D);
	}

	@Override
	protected void attackEntity(Entity par1Entity, float par2)
	{
		if((attackTime <= 0) && (par2 < 2.0F) && (par1Entity.boundingBox.maxY > boundingBox.minY) && (par1Entity.boundingBox.minY < boundingBox.maxY))
		{
			attackTime = 20;
			attackEntityAsMob(par1Entity);
		}
	}

	@Override
	public boolean attackEntityAsMob(Entity par1Entity)
	{
		float f = (float)getEntityAttribute(SharedMonsterAttributes.attackDamage).getAttributeValue();
		int i = 0;

		if(par1Entity instanceof EntityLivingBase)
		{
			f += EnchantmentHelper.getEnchantmentModifierLiving(this, (EntityLivingBase)par1Entity);
			i += EnchantmentHelper.getKnockbackModifier(this, (EntityLivingBase)par1Entity);
		}

		boolean flag = par1Entity.attackEntityFrom(DamageSource.causeMobDamage(this), f);

		if(flag)
		{
			if(i > 0)
			{
				par1Entity.addVelocity(-MathHelper.sin((rotationYaw * (float)Math.PI) / 180.0F) * i * 0.5F, 0.1D, MathHelper.cos((rotationYaw * (float)Math.PI) / 180.0F) * i * 0.5F);
				motionX *= 0.6D;
				motionZ *= 0.6D;
			}

			int j = EnchantmentHelper.getFireAspectModifier(this);

			if(j > 0)
			{
				par1Entity.setFire(j * 4);
			}

			if(par1Entity instanceof EntityLivingBase)
			{
				EnchantmentThorns.func_92096_a(this, (EntityLivingBase)par1Entity, rand);
			}
		}

		return flag;
	}

	@Override
	public boolean attackEntityFrom(DamageSource par1DamageSource, float par2)
	{
		if(isEntityInvulnerable()) return false;
		else if(super.attackEntityFrom(par1DamageSource, par2))
		{
			Entity entity = par1DamageSource.getEntity();

			if((riddenByEntity != entity) && (ridingEntity != entity))
			{
				if(entity != this)
				{
					entityToAttack = entity;
				}

				return true;
			}
			else return true;
		}
		else return false;
	}

	@Override
	protected boolean canDespawn()
	{
		return false;
	}

	@Override
	public EntityAgeable createChild(EntityAgeable entityAgeable)
	{
		return new EntityCamper(worldObj);
	}

	@Override
	protected void entityInit()
	{
		super.entityInit();
		dataWatcher.addObject(16, Integer.valueOf(0));
	}

	@Override
	public void func_110297_a_(ItemStack stack)
	{
		if(!worldObj.isRemote && (livingSoundTime > (-getTalkInterval() + 20)))
		{
			livingSoundTime = -getTalkInterval();

			if(stack != null)
			{
				playSound("mob.villager.yes", getSoundVolume(), getSoundPitch());
			}
			else
			{
				playSound("mob.villager.no", getSoundVolume(), getSoundPitch());
			}
		}
	}

	public void func_82187_q()
	{}

	@SideOnly(Side.CLIENT)
	private void generateRandomParticles(String par1Str)
	{
		for(int i = 0; i < 5; ++i)
		{
			double d0 = rand.nextGaussian() * 0.02D;
			double d1 = rand.nextGaussian() * 0.02D;
			double d2 = rand.nextGaussian() * 0.02D;
			worldObj.spawnParticle(par1Str, (posX + (rand.nextFloat() * width * 2.0F)) - width, posY + 1.0D + (rand.nextFloat() * height), (posZ + (rand.nextFloat() * width * 2.0F)) - width, d0, d1, d2);
		}
	}

	@Override
	public EntityPlayer getCustomer()
	{
		return buyingPlayer;
	}

	@Override
	protected String getDeathSound()
	{
		return "mob.villager.death";
	}

	public int getGender()
	{
		return dataWatcher.getWatchableObjectInt(16);
	}

	@Override
	protected String getHurtSound()
	{
		return "mob.villager.hit";
	}

	@Override
	protected String getLivingSound()
	{
		return isTrading()? "mob.villager.haggle":"mob.villager.idle";
	}

	@Override
	public MerchantRecipeList getRecipes(EntityPlayer player)
	{
		if(buyingList == null)
		{
			addDefaultEquipmentAndRecipies(1);
		}

		return buyingList;
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void handleHealthUpdate(byte par1)
	{
		if(par1 == 12)
		{
			generateRandomParticles("heart");
		}
		else if(par1 == 13)
		{
			generateRandomParticles("angryVillager");
		}
		else if(par1 == 14)
		{
			generateRandomParticles("happyVillager");
		}
		else
		{
			super.handleHealthUpdate(par1);
		}
	}

	@Override
	public boolean interact(EntityPlayer player)
	{
		ModAchievements.encounterCamper.addStatToPlayer(player);

		ItemStack itemstack = player.inventory.getCurrentItem();
		boolean flag = (itemstack != null) && (itemstack.itemID == Item.monsterPlacer.itemID);

		if(!flag && isEntityAlive() && !isTrading() && !isChild() && !player.isSneaking())
		{
			if(!worldObj.isRemote)
			{
				setCustomer(player);
				player.displayGUIMerchant(this, "Camper");
			}

			return true;
		}
		else return super.interact(player);
	}

	@Override
	public boolean isAIEnabled()
	{
		return true;
	}

	public boolean isMating()
	{
		return isMating;
	}

	public boolean isPlaying()
	{
		return isPlaying;
	}

	public boolean isTrading()
	{
		return buyingPlayer != null;
	}

	@Override
	public void readEntityFromNBT(NBTTagCompound tag)
	{
		super.readEntityFromNBT(tag);
		setGender(tag.getInteger("gender"));
		wealth = tag.getInteger("Riches");
		theCampfire = tag.getIntArray("theCampfire");

		if(tag.hasKey("Offers"))
		{
			NBTTagCompound nbttagcompound1 = tag.getCompoundTag("Offers");
			buyingList = new MerchantRecipeList(nbttagcompound1);
		}
	}

	@Override
	public void setCustomer(EntityPlayer player)
	{
		buyingPlayer = player;
	}

	public void setGender(int par1)
	{
		dataWatcher.updateObject(16, Integer.valueOf(par1));
	}

	public void setMating(boolean par1)
	{
		isMating = par1;
	}

	public void setPlaying(boolean par1)
	{
		isPlaying = par1;
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void setRecipes(MerchantRecipeList merchantRecipeList)
	{}

	@Override
	protected void updateAITick()
	{
		if((theCampfire == null) || (theCampfire.length < 3))
		{
			theCampfire = new int[]{0, 0, 0};
		}

		if((updateHome-- <= 0) && (worldObj.getBlockId(theCampfire[0], theCampfire[1], theCampfire[2]) != ModBlocks.campfireBase.blockID))
		{
			setHomeArea(theCampfire[0], theCampfire[1], theCampfire[2], -1);

			Iterator tiles = worldObj.getChunkFromBlockCoords((int)posX, (int)posZ).chunkTileEntityMap.values().iterator();

			while(tiles.hasNext())
			{
				TileEntity tile = (TileEntity)tiles.next();

				if(tile instanceof TileEntityCampfireCook)
				{
					theCampfire = new int[]{tile.xCoord, tile.yCoord, tile.zCoord};
					setHomeArea(theCampfire[0], theCampfire[1], theCampfire[2], 8);
					break;
				}
			}
			updateHome = 100;
		}
		else if(updateHome == 20)
		{
			setHomeArea(theCampfire[0], theCampfire[1], theCampfire[2], 8);
		}

		if(!isTrading() && (timeUntilReset > 0))
		{
			--timeUntilReset;

			if(timeUntilReset <= 0)
			{
				if(needsInitilization)
				{
					if(buyingList.size() > 1)
					{
						Iterator iterator = buyingList.iterator();

						while(iterator.hasNext())
						{
							MerchantRecipe merchantrecipe = (MerchantRecipe)iterator.next();

							if(merchantrecipe.func_82784_g())
							{
								merchantrecipe.func_82783_a(rand.nextInt(6) + rand.nextInt(6) + 2);
							}
						}
					}

					addDefaultEquipmentAndRecipies(1);
					needsInitilization = false;
				}

				addPotionEffect(new PotionEffect(Potion.regeneration.id, 200, 0));
			}
		}

		super.updateAITick();
	}

	@Override
	public void useRecipe(MerchantRecipe merchantRecipe)
	{
		merchantRecipe.incrementToolUses();
		livingSoundTime = -getTalkInterval();
		playSound("mob.villager.yes", getSoundVolume(), getSoundPitch());

		if(merchantRecipe.hasSameIDsAs((MerchantRecipe)buyingList.get(buyingList.size() - 1)))
		{
			timeUntilReset = 40;
			needsInitilization = true;

			if(buyingPlayer != null)
			{
				lastBuyingPlayer = buyingPlayer.getCommandSenderName();
			}
			else
			{
				lastBuyingPlayer = null;
			}
		}

		if(merchantRecipe.getItemToBuy().itemID == Item.emerald.itemID)
		{
			wealth += merchantRecipe.getItemToBuy().stackSize;
		}
	}

	@Override
	public void writeEntityToNBT(NBTTagCompound tag)
	{
		super.writeEntityToNBT(tag);
		tag.setInteger("gender", getGender());
		tag.setInteger("Riches", wealth);
		tag.setIntArray("theCampfire", theCampfire);

		if(buyingList != null)
		{
			tag.setCompoundTag("Offers", buyingList.getRecipiesAsTags());
		}
	}
}

File: scala/com/rikmuld/camping/features/entities/camper/EntityCamper.scala
package com.rikmuld.camping.features.entities.camper

import java.util.{ArrayList, Random}

import com.rikmuld.camping.CampingMod
import com.rikmuld.camping.Definitions.Parts
import com.rikmuld.camping.Library._
import com.rikmuld.camping.features.entities.bear.EntityBear
import com.rikmuld.corerm.advancements.TriggerHelper
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity._
import net.minecraft.entity.ai._
import net.minecraft.entity.monster.EntityMob
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.{Items, SoundEvents}
import net.minecraft.item.{Item, ItemAxe, ItemStack}
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.datasync.{DataParameter, DataSerializers, EntityDataManager}
import net.minecraft.pathfinding.PathNavigateGround
import net.minecraft.util._
import net.minecraft.util.math.{BlockPos, MathHelper}
import net.minecraft.util.text.TextComponentString
import net.minecraft.village.{MerchantRecipe, MerchantRecipeList}
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

import scala.collection.mutable.HashMap

object EntityCamper {
  final val GENDER:DataParameter[Integer] = EntityDataManager.createKey(classOf[EntityCamper], DataSerializers.VARINT);

  val recipeListRaw = new HashMap[Item, Tuple[Integer, Integer]]()

  recipeListRaw(CampingMod.OBJ.venisonCooked) = new Tuple(Integer.valueOf(-4), Integer.valueOf(1))
  recipeListRaw(CampingMod.OBJ.knife) = new Tuple(Integer.valueOf(1), Integer.valueOf(2))
  recipeListRaw(CampingMod.OBJ.backpack) = new Tuple(Integer.valueOf(1), Integer.valueOf(2))
  recipeListRaw(CampingMod.OBJ.furLeg) = new Tuple(Integer.valueOf(4), Integer.valueOf(6))
  recipeListRaw(CampingMod.OBJ.furHead) = new Tuple(Integer.valueOf(3), Integer.valueOf(5))
  recipeListRaw(CampingMod.OBJ.furChest) = new Tuple(Integer.valueOf(5), Integer.valueOf(8))
  recipeListRaw(CampingMod.OBJ.furBoot) = new Tuple(Integer.valueOf(2), Integer.valueOf(4))
  recipeListRaw(CampingMod.OBJ.parts) = new Tuple(Integer.valueOf(1), Integer.valueOf(2))
}

class EntityCamper(worldIn: World) extends EntityCreature(worldIn) with IMerchant with INpc {
  setGender(rand.nextInt(2))
  setSize(0.6F, 1.8F)
  getNavigator.asInstanceOf[PathNavigateGround].setCanSwim(true)
  tasks.addTask(1, new EntityAISwimming(this))
  tasks.addTask(2, new EntityAIAttackMelee(this, 1.0F, false))
  tasks.addTask(4, new EntityAIWatchClosest2(this, classOf[EntityPlayer], 3.0F, 1.0F))
  tasks.addTask(6, new EntityAIWander(this, 0.6D))
  targetTasks.addTask(0, new EntityAINearestAttackableTarget(this, classOf[EntityMob], false))
  targetTasks.addTask(0, new EntityAINearestAttackableTarget(this, classOf[EntityBear], false))

  var playerBuy: EntityPlayer = null
  var recipeList: MerchantRecipeList = _
  var camp:Option[Campsite] = None

  override def getWorld: World = world
  def setGender(gender: Int) = dataManager.set(EntityCamper.GENDER, Integer.valueOf(gender))
  def getGender: Int = dataManager.get(EntityCamper.GENDER)
  override def getPos: BlockPos = new BlockPos(this)
  override def writeEntityToNBT(tag: NBTTagCompound) {
    super.writeEntityToNBT(tag)
    tag.setInteger("gender", getGender)
    camp.map { campsite =>
      tag.setBoolean("hasCamp", true)
      tag.setTag("campsite", campsite.toNBT(new NBTTagCompound))
    }
  }
  override def readEntityFromNBT(tag: NBTTagCompound) {
    super.writeEntityToNBT(tag)
    setGender(tag.getInteger("gender"))
    if(tag.hasKey("hasCamp")) camp = Some(Campsite.fromNBT(this, tag.getCompoundTag("campsite")))
  }
  protected override def applyEntityAttributes {
    super.applyEntityAttributes();
    getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.4D)
    getAttributeMap().registerAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).setBaseValue(4.0D)
  }
  override def canDespawn = false
  override def entityInit {
    super.entityInit
    dataManager.register(EntityCamper.GENDER, Integer.valueOf(0))
  }
  override def onUpdate {
    super.onUpdate
    camp.map {campsite => campsite.update }
  }
  protected override def getDeathSound:SoundEvent = SoundEvents.ENTITY_VILLAGER_DEATH
  protected override def getHurtSound(source: DamageSource):SoundEvent = SoundEvents.ENTITY_VILLAGER_HURT
  protected override def getAmbientSound:SoundEvent = SoundEvents.ENTITY_VILLAGER_AMBIENT
  override def attackEntityAsMob(entity: Entity): Boolean = {
    var f = this.getEntityAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).getAttributeValue.toFloat
    var i = 0
    if (entity.isInstanceOf[EntityLivingBase]) {
      f += EnchantmentHelper.getModifierForCreature(this.getHeldItemMainhand(), entity.asInstanceOf[EntityLivingBase].getCreatureAttribute());
      i += EnchantmentHelper.getKnockbackModifier(this);
    }
    val flag = entity.attackEntityFrom(DamageSource.causeMobDamage(this), f)
    if (flag) {
      if (i > 0) {
        entity.addVelocity((-MathHelper.sin(this.rotationYaw * Math.PI.toFloat / 180.0F) * i.toFloat * 0.5F).toDouble, 0.1D, (MathHelper.cos(this.rotationYaw * Math.PI.toFloat / 180.0F) * i.toFloat * 0.5F).toDouble)
        this.motionX *= 0.6D
        this.motionZ *= 0.6D
      }
      val j = EnchantmentHelper.getFireAspectModifier(this)
      if (entity.isInstanceOf[EntityPlayer]) {
          val player = entity.asInstanceOf[EntityPlayer]
          val itemstack = this.getHeldItemMainhand();
          val itemstack1 = if(player.isHandActive()) player.inventory.getCurrentItem else null;

          if (itemstack != null && itemstack1 != null && itemstack.getItem.isInstanceOf[ItemAxe] && itemstack1.getItem() == Items.SHIELD) {
              val f1 = 0.25F + EnchantmentHelper.getEfficiencyModifier(this) * 0.05F;

              if (this.rand.nextFloat() < f1)
              {
                  player.getCooldownTracker().setCooldown(Items.SHIELD, 100);
                  this.world.setEntityState(player, 30.toByte);
              }
          }
      }
      if (j > 0) entity.setFire(j * 4)
      this.applyEnchantments(this, entity)
    }
    flag
  }
  override def processInteract(player: EntityPlayer, hand:EnumHand): Boolean = {
    val itemstack = player.inventory.getCurrentItem
    val flag = (itemstack != null) && (itemstack.getItem == Items.SPAWN_EGG)
    if (!flag && isEntityAlive && !isTrading && !player.isSneaking) {
      if (!world.isRemote) {
        setCustomer(player)
        player.displayVillagerTradeGui(this)
        TriggerHelper.trigger(AdvancementInfo.CAMPER_INTERACT, player, this)
      }
      true
    } else super.processInteract(player, hand)
  }
  override def setCustomer(player: EntityPlayer) = playerBuy = player
  override def getCustomer(): EntityPlayer = playerBuy
  override def getRecipes(player: EntityPlayer): MerchantRecipeList = if (recipeList == null) setRecipeList else recipeList
  @SideOnly(Side.CLIENT)
  override def setRecipes(recipes: MerchantRecipeList) {}
  override def useRecipe(resipes: MerchantRecipe) {}
  override def verifySellingItem(stack:ItemStack) {}
  def isTrading: Boolean = getCustomer != null
  def setRecipeList: MerchantRecipeList = {
    recipeList = new MerchantRecipeList()
    if (getGender == 0) {
      addBlacksmithItem(recipeList, CampingMod.OBJ.furBoot, 0, rand, 0.4F)
      addBlacksmithItem(recipeList, CampingMod.OBJ.furChest, 0, rand, 0.3F)
      addBlacksmithItem(recipeList, CampingMod.OBJ.furHead, 0, rand, 0.4F)
      addBlacksmithItem(recipeList, CampingMod.OBJ.furLeg, 0, rand, 0.4F)
      addBlacksmithItem(recipeList, CampingMod.OBJ.knife, 0, rand, 0.6F)
      addBlacksmithItem(recipeList, CampingMod.OBJ.backpack, 0, rand, 0.5F)
    } else {
      addBlacksmithItem(recipeList, CampingMod.OBJ.venisonCooked, 0, rand, 0.6F)
      addBlacksmithItem(recipeList, CampingMod.OBJ.parts, Parts.PAN, rand, 0.6F)
    }
    if (recipeList.isEmpty()) {
      addBlacksmithItem(recipeList, CampingMod.OBJ.knife, 0, rand, 0.6F)
    }
    recipeList
  }
  override def getDisplayName = new TextComponentString("Camper")
  def addBlacksmithItem(merchantRecipeList: MerchantRecipeList, item: Item, meta: Int, random: Random, par3: Float) {
    if (random.nextFloat() < par3) {
      val j = randomCount(item, random)
      var itemstack: ItemStack = null
      var itemstack1: ItemStack = null
      if (j < 0) {
        itemstack = new ItemStack(Items.EMERALD, 1, 0)
        itemstack1 = new ItemStack(item, -j, meta)
      } else {
        itemstack = new ItemStack(Items.EMERALD, j, 0)
        itemstack1 = new ItemStack(item, 1, meta)
      }
      merchantRecipeList.asInstanceOf[ArrayList[MerchantRecipe]].add(new MerchantRecipe(itemstack, itemstack1))
    }
  }
  private def randomCount(item: Item, random: Random): Int = {
    val tuple = EntityCamper.recipeListRaw(item).asInstanceOf[Tuple[Integer, Integer]]
    if (tuple == null) 1 else (if (tuple.getFirst.asInstanceOf[java.lang.Integer].intValue() >= tuple.getSecond.asInstanceOf[java.lang.Integer].intValue()) tuple.getFirst.asInstanceOf[java.lang.Integer].intValue() else tuple.getFirst.asInstanceOf[java.lang.Integer].intValue() + random.nextInt(tuple.getSecond.asInstanceOf[java.lang.Integer].intValue() - tuple.getFirst.asInstanceOf[java.lang.Integer].intValue()))
  }
  def setCampsite(campsite:Option[Campsite]) = camp = campsite
}